### ISSUE TITLE
wasm32-unknown-unknown without JS bindgen

### ISSUE THREAD

**Issue by kkonevets** [rust-random/getrandom#340]
I am posting this issue also here, because it is related to `getrandom`. Please comment on.
https://github.com/RustCrypto/nacl-compat/issues/90

TLDR:
Can I somehow get rid of `js` feature but still be able to compile to `wasm32-unknown-unknown`?

**Comment by josephlr** [comment]
Yes, as noted in the docs, you can enable the `"custom"` feature and provide your own RNG implementation.

**Comment by kkonevets** [comment]
> Yes, as noted in the docs, you can enable the `"custom"` feature and provide your own RNG implementation.

I have seen that, but I don't know how to do that. Is there some example of using custom RNG  implementation?

**Comment by josephlr** [comment]
> > Yes, as noted in the docs, you can enable the `"custom"` feature and provide your own RNG implementation.
> 
> I have seen that, but I don't know how to do that. Is there some example of using custom RNG implementation?

Is there part of the documentation for this crate that is unclear?

**Comment by kkonevets** [comment]
> > TLDR: Can I somehow get rid of `js` feature
> 
> Also the `"js"` feature is off by default, so it's mostly about not enabling that feature.

But `js` feature is the only way to  target `wasm32-unknown-unknown`, the defaults don't work

**Comment by josephlr** [comment]
> But `js` feature is the only way to target `wasm32-unknown-unknown`

This is incorrect. You have to enable _some_ implementation in order for `wasm32-unknown-unknown` to work.
  - If you enable the `js` feature, you will use the JavaScript bindgen implementation
  - If you enable the `custom` feature, you will use your own implementation provided via [`register_custom_getrandom`](https://docs.rs/getrandom/latest/getrandom/macro.register_custom_getrandom.html)
  - If you don't enable either feature, this crate will not compile when targeting `wasm32-unknown-unknown`

**Comment by kkonevets** [comment]
I have just enabled `custom` feature and all my tests are passing. Any idea why? I use getrandom with [crypto-box](https://github.com/RustCrypto/nacl-compat/tree/master/crypto_box). Does it mean that `crypto-box` does not use getrandom implementation? You know, I generate lots of random public, secret keys and nonces. I am not a cryptographer, sorry.

**Comment by josephlr** [comment]
> I have just enabled `custom` feature and all my tests are passing. Any idea why?

I would make sure that you are:
  - Actually targeting `wasm32-unknown-unknown`
  - Actually generating random keys via `getrandom`

You should get a linker if error if you try to call the custom implementation and one is not registered.

**Comment by kkonevets** [comment]
I am sorry, it works for `std`, but fails in `wasm32-unknown-unknown` with

```
Uncaught (in promise) LinkError: WebAssembly.instantiate(): Import #0 module="env" function="__getrandom_custom" error: function import requires a callable
await (async)
(anonymous) @ (index):55
```
Is there any usable implementation of custom RNG other then `use failure_getrandom::always_fail`? The point is that I want `getrandom` to give me randomness, but now it does not.

**Comment by josephlr** [comment]
That linker error is what I would have expected. You need to use [`register_custom_getrandom`](https://docs.rs/getrandom/latest/getrandom/macro.register_custom_getrandom.html) to fix that error.

> Is there any usable implementation of custom RNG other then `use failure_getrandom::always_fail`? 

I can't answer this question, as it depends on what platform you are targeting. You will need to write your own implementation to get random numbers.

**Comment by kkonevets** [comment]
Thanks! I got it.
For those who will read this issue, here is what you should do.

```rust
use getrandom::{self, register_custom_getrandom};

extern "C" {
	// call host OS random number generator
	fn random(dest: *mut c_char, len: usize);
}

pub fn getrandom_custom(dest: &mut [u8]) -> Result<(), getrandom::Error> {
        // this can actually fail, so you should return an appropriate error
	unsafe { random(dest.as_mut_ptr() as *mut i8, dest.len()) };
	Ok(())
}

register_custom_getrandom!(getrandom_custom);
```

If you use wasmer, then you should create an import object with FunctionEnv that calls host OS random number generator
